spec/unit/pops/types/p_object_type_spec.rb

```diff
@@ -1136,6 +1140,282 @@ def compile(code)
+          },
+          attributes => {
+            p1 => String,
+            p2 => String
+          }]
+        $x = MyType('hello', 'world')
+        notice($x =~ MyType)
+        notice($x =~ MyType[default])
+        notice($x =~ MyType[default, default])
+        notice($x =~ MyType[default, undef])
+        PUPPET
+      end
+
+      it 'Required parameters cannot be set using default' do
+        expect{eval_and_collect_notices(<<-PUPPET, node)}.to raise_error(/No value provided for required type_parameter MyType\[p1\]/)
+        type MyType = Object[
+          type_parameters => {
+            p1 => { type => Variant[Undef,String,Regexp,Type[Enum],Type[Pattern]] },
+            p2 => { type => Variant[Undef,String,Regexp,Type[Enum],Type[Pattern]] },
+          },
+          attributes => {
+            p1 => String,
+            p2 => String
+          }]
+        $x = MyType('hello', 'world')
+        notice($x =~ MyType)
+        notice($x =~ MyType[default, default])
+        PUPPET
+      end
+
+      it 'Required parameters must be provided' do
+        expect{eval_and_collect_notices(<<-PUPPET, node)}.to raise_error(/No value provided for required type_parameter MyType\[p2\]/)
+        type MyType = Object[
+          type_parameters => {
+            p1 => { type => Variant[Undef,String,Regexp,Type[Enum],Type[Pattern]] },
+            p2 => { type => Variant[Undef,String,Regexp,Type[Enum],Type[Pattern]] },
+          },
+          attributes => {
+            p1 => String,
+            p2 => String
+          }]
+        $x = MyType('hello', 'world')
+        notice($x =~ MyType)
+        notice($x =~ MyType['hello'])
+        PUPPET
+      end
+
+      it 'Required parameters can be set using undef' do
+        expect(eval_and_collect_notices(<<-PUPPET, node)).to eql(['true', 'false', 'true', 'true'])
+        type MyType = Object[
+          type_parameters => {
+            p1 => { type => Variant[Undef,String,Regexp,Type[Enum],Type[Pattern]] },
+            p2 => { type => Variant[Undef,String,Regexp,Type[Enum],Type[Pattern]] },
+          },
+          attributes => {
+            p1 => { type => Optional[String], value => undef },
+            p2 => { type => Optional[String], value => undef }
+          }]
+        $x = MyType('hello')
+        notice($x =~ MyType)
+        notice($x =~ MyType[undef, undef])
+        notice($x =~ MyType[/hello/, undef])
+        notice(MyType() =~ MyType[undef, undef])
+        PUPPET
+      end
+
+      it 'default values are considered when computing assignability' do
+        expect(eval_and_collect_notices(<<-PUPPET, node)).to eql(['true', 'true', 'true', 'false'])
+        type MyType = Object[
+          type_parameters => {
+            p1 => { type => String, value => 'hello' },
+            p2 => { type => String, value => 'world' },
+          }]
+        $x = MyType['hello']
+        notice($x <= MyType)
+        notice($x <= MyType['hello'])
```

Comment:
This is surprising (or the test title is off) - the instance has `p1 = 'hello'` and `p2 = undef`, how can that match `MyType['hello']` and consider the default p2 to match `world` when p2 attribute is `undef`? 
While this line (1393) is less surprising (no visible constraint on p2), the next is more so since it is explicit that p2 must match `world` but the instance does not have this attribute value. ??? wtf ??

Reply:
Not sure I follow. Both type parameters have a default here. No {{undef}} is ever compared. I.e.
{code:puppet}
MyType['hello'] == MyType['hello', 'world'] # because 'world' is the default.
{code}
that gives that all tests except the last one evaluates to {{true}}. The last one changes the first parameter to 'my' and it doesn't match the 'hello' of $x.

Reply:
I think the conceptual problem here is that `T` is unconstrained whereas `T[something]` makes all constraints kick in with their defaults when the defaults makes the type more constrained than "unconstrained". IIRC all defaults so far have had values that match "unconstrained" (e.g. min-integer, max-integer) so we have never seen those (surprising) cases earlier. We had the opposite though - Hash was more constrained, and you could unconstrained via parametrization.

Reply:
We discussed this at length in our last meeting since it was a blocker for me and concluded that a completely unconstrained type (specified without parameters) is different from a type constrained with its parameter defaults. This implementation is based on that assumption. If we decided that a completely unconstrained type is exactly the same as a type constrained by its defaults, then it is back to the drawing board. That would imply that any type declared with a type parameter that has defaults must take those defaults into account at all times.

Reply:
works if you are not allowed to have defaults that constrain

Reply:
What would be the point in having defaults if they don't impose any constraints? What's the difference between that and saying that the type parameter is unset?

Reply:
It may in fact be of no value to have the defaults for what we are supporting with the base implementation (compare against attributes). The only reasonable default is always the type of the attribute == the same as no additional constraint == the same as not set. It is only when there is some other mapping than "type param x matches against attribute x" that defaults start to become meaningful (for example `Integer[$min, $max]` since you don't know what they values are and you want to parameterize with variables - but then I question that the default values when used should add constraints; i.e. they are just there to be able to assign a value to a variable that means "do not constrain in this dimension". I guess I don't see the value in being able to have other defaults than unconstrained/unset - the examples where defaults are used are just confusing IMO.

Reply:
There seem to be two solutions to the dilemma.

1. Defaults can be used (a default can be any value) and `T[default, default, default,...] == T`. This implies that a comparison against the base type will take defaults into account. It also means that a type that uses only the defaults is represented by the base type, not a type extension.
2. Defaults cannot be used at all. A type parameter value is either unset or explicitly given a value.

I'm fine either way. What's your preference?

Reply:
Preference is (2) - with this motivation:
Since sanity means that (1) requires that no default may impose a constraint that breaks the rule "unparameterized T means all instances" it then also becomes somewhat useless for the use case we now want to support. Defaults may be needed in the future, it may be meaningful in a derived type for example (but I cannot come up with complete use cases). So... let's implement (2) since that is both sane and does not block us from adding this in the future - I prefer that to introducing something which has hard to understand implications.